proc/cut.go

```diff
@@ -1,74 +1,109 @@
+	return &Cut{
+		Base:   Base{Context: c, Parent: parent},
+		fields: fields,
+		cutmap: make(map[int]*zson.Descriptor),
+	}
 }
 
-// Check the bitmap of fields that we've seen.  If a requested field
-// never appeared in the input, emit a warning about it.  This should be
-// called once when this proc reaches the end of the stream.
-func (c *Cut) WarnUnseen() {
-	if !c.sawRecord {
-		return
+func (c *Cut) lookup(in *zson.Descriptor) *zson.Descriptor {
+	d, ok := c.cutmap[in.ID]
+	if ok {
+		return d
 	}
-	for i, name := range c.fields {
-		if (c.seenFields & (1 << i)) == 0 {
-			c.Warnings <- fmt.Sprintf("Cut field %s not present in input", name)
+	var columns []zeek.Column
+	for _, field := range c.fields {
+		colno, ok := in.ColumnOfField(field)
+		if !ok {
+			// a field is missing... block this descriptor
+			c.cutmap[in.ID] = nil
+			c.nblocked++
+			return nil
+		}
+		columns = append(columns, in.Type.Columns[colno])
+	}
+	out := c.Resolver.GetByColumns(columns)
+	c.cutmap[in.ID] = out
+	return out
+}
+
+// CreateCut returns a new record value derived by keeping only the fields
+// specified by name in the fields slice.
+func (c *Cut) cut(d *zson.Descriptor, in *zson.Record) (*zson.Record, error) {
+	var zv zval.Encoding
+	for _, column := range d.Type.Columns {
+		colno, ok := in.ColumnOfField(column.Name)
+		if !ok {
+			return nil, ErrNoField
 		}
+		zv = zval.Append(zv, in.Slice(colno), zeek.IsContainerType(column.Type))
 	}
+	return zson.NewRecordNoTs(d, zv), nil
+}
+
+func (c *Cut) warn() {
+	if len(c.cutmap) > c.nblocked {
+		return
+	}
+	flds := strings.Join(c.fields, ",")
+	plural := ""
+	msg := "not present in input"
+	if len(c.fields) > 1 {
+		plural = "s"
+		msg = "not present together in inout"
+	}
+	c.Warnings <- fmt.Sprintf("Cut field%s %s %s", plural, flds, msg)
 }
 
 func (c *Cut) Pull() (zson.Batch, error) {
 	batch, err := c.Get()
 	if EOS(batch, err) {
-		c.WarnUnseen()
+		c.warn()
 		return nil, err
 	}
 	defer batch.Unref()
-	c.sawRecord = true
 	//
 	// Make new records with only the fields specified.
 	// If a field specified doesn't exist, we don't include that record.
 	// if the types change for the fields specified, we drop those records.
 	//
-	out := make([]*zson.Record, 0, batch.Length())
-	var d *zson.Descriptor
+	recs := make([]*zson.Record, 0, batch.Length())
 	for k := 0; k < batch.Length(); k++ {
 		in := batch.Index(k)
-		r, found, err := c.Resolver.CreateCut(in, c.fields)
-		c.seenFields |= found
+		d := c.lookup(in.Descriptor)
+		if d == nil {
+			continue
+		}
+		out, err := c.cut(d, in)
 		if err != nil {
-			c.WarnUnseen()
 			return nil, err
 		}
-		if r != nil {
-			if d == nil {
-				d = r.Descriptor
-			}
-			// Check that the types are the same throughout
-			// if any of the types change, we ignore those records.
-			if d == r.Descriptor {
-				out = append(out, r)
-			}
-		}
+		recs = append(recs, out)
 	}
-	if d == nil {
-		c.WarnUnseen()
+	if len(recs) == 0 {
+		c.warn()
```

Comment:
I know this was already here but feels like this shouldn't warn on every batch that doesn't have a match.

Reply:
This case is at EOS.

Reply:
Right.